app/src/trackPlayerContext.tsx

```diff
@@ -28,61 +65,285 @@ export function usePlayer() {
-  const state = await RNTrackPlayer.getState();
-  var track: Track = null;
-  if (currentIndex != null) {
-    track = await RNTrackPlayer.getTrack(currentIndex);
-  }
-
-  return {
-    currentTrack: track,
-    currentIndex: currentIndex,
-    recordsCount: queue.length,
-    isPlaying: state == State.Playing,
-  };
-}
+const QUEUE_STORAGE_KEY = "audio_queue";
 
 const PlayerContextProvider: FC = ({ children }) => {
   const [state, _setState] = useState<PlayerState>(contextDefaultValues.state);
+  const [progress, setProgress] = useState<TrackProgress>(
+    contextDefaultValues.progress
+  );
+  const playerProgress = useProgress();
 
-  useTrackPlayerEvents(
-    [Event.PlaybackState, Event.PlaybackTrackChanged, Event.PlaybackError],
-    async (event) => {
-      const newState = await createPlayerState();
-      setState(newState);
-
-      if (event.type == Event.PlaybackTrackChanged) {
-        const prevTrackIndex = event.track;
-        if (prevTrackIndex != null) {
-          const prevTrack = await RNTrackPlayer.getTrack(prevTrackIndex);
-          const prevTrackPosition = event.position;
-          if (
-            prevTrack != null &&
-            prevTrackPosition != null &&
-            prevTrackPosition > 5
-          ) {
-            logArticlePlayed(prevTrack.id);
-          }
+  // init
+  useEffect(() => {
+    getSavedState().then((savedState) => {
+      if (savedState != null) {
+        console.info("Loading saved state.");
+        const notPlayingState = {
+          ...savedState,
+          isPlaying: false,
+        };
+        _setState(notPlayingState);
+        if (notPlayingState.currentTrack != null) {
+          playerInstance.setActiveTrack(notPlayingState.currentTrack);
         }
-        const nextTrack = event.nextTrack;
       }
+    });
+  }, []);
+
+  useEffect(() => {
+    setProgress(playerProgress);
+
+    if (playerProgress.position > 5 && state.isPlaying) {
+      reportCurrentTrackPlayback();
     }
-  );
+  }, [playerProgress]);
+
+  useEffect(() => {
+    console.info("State changed");
+    saveState();
+  }, [state]);
+
+  /**
+   * State persisting
+   */
+  const getSavedState = async () => {
+    try {
+      const jsonData = await AsyncStorage.getItem(QUEUE_STORAGE_KEY);
+      if (jsonData != null) {
+        const stateObject: PlayerState = JSON.parse(jsonData);
+        return stateObject;
+      }
+      return null;
+    } catch (e) {
+      console.error(e);
+      return null;
+    }
+  };
+
+  const saveState = async () => {
+    try {
+      const jsonValue = JSON.stringify(state);
+      await AsyncStorage.setItem(QUEUE_STORAGE_KEY, jsonValue);
+    } catch (e) {
+      console.error(e);
+    }
+  };
+
+  /**
+   * Player events handling
+   */
+
+  useTrackPlayerEvents([Event.PlaybackQueueEnded], async (event) => {
+    console.info("PlaybackQueueEnded", event, progress);
+    const _state = {
+      ...state,
+      isPlaying: false,
+    };
+    _setState(_state);
```

Comment:
_setState má (méně známou) variantu jako "reducer" funkci, která se hodí v případě, že modifikujeme existující state... má tu výhodu, že nejsme závislí na rerenderu, abychom měli aktuální state, a zároveň setState pak nemá dependenci na state, což řeší problém s updaty v useEffect

```suggestion
    _setState(state => ({ ...state, isPlaying: false }));
```